### PR TITLE
Implement `autobib path`

### PR DESCRIPTION
**Pull request by alexrutar**
A place to work on the `autobib path` features.

Checklist:
- [x] Implement basic path logic.
- [x] ~Add commands to insert new files.~
- [x] ~Add commands to list all files.~ I think this should not be implemented here, since there are too many ways to 'list files' (use `find`? `walkdir`? `ls`?) Instead, I think this should be part of `autobib find`, or similar. See #77 
- [x] ~Figure out how to unify the implementation logic (so that the "path" initialization can be handled *before* we start processing the subcommand).~
- [x] ~Implement filesystem locking logic so we do not have race conditions on edit (malicious race conditions notwithstanding, but we should mainly care about the "multiple instances of autobib using the same database" situation).~

### CODE REVIEW

**Comment by alexrutar** [comment]
Having thought about this more, I am leaning towards this being a separate tool. There are so many choices as to how files should be organized within the relevant directory and I feel like any implementation is implicitly making a lot of choices about this on behalf of the user.

For example, adding a super crude `autobib path add` as a shell script is as easy as
```fish
cp -i path/to/source (autobib path <key>)/target
```

I think it would be best to add all of the 'plumbing' tools required to make this convenient. @wupr What do you think are the relevant tools which we should implement?

1. Canonical `path` locations definitely counts (and is already implemented).
2. File renaming support of some form would be great. But this could be a separate tool, like `autobib format <KEY> <FORMAT_SPEC>` which takes a key and a specifier like `%(author)%(year)` or whatever, and then prints the corresponding rename rule to the command line. We could even add options like `--path-friendly` which will intentionally normalize the file to be 'path-like', for usage with `autobib path`.
3. File searching is (in my opinion) quite canonical, since this does not depend on the specific organization of the files.

On the other hand, I do believe that it is nice to manage such things in some sort of canonical format on behalf of the user, since it makes the tool easier to use. Otherwise everyone needs to find the relevant script, download a separate tool, or whatever. I am curious what your thoughts about this are.

File: src/record/key.rs [hunk 1]
```diff
@@ -217,6 +223,44 @@ impl RemoteId {
...
+    /// provider/xx/xx/xx/base32-encoding-of-sub-id
+    /// ```
+    /// The 30 bit header is formed by converting the u64 output of the
+    /// [highway](https://crates.io/crates/highway) hashing algorithm applied to the sub-id into
+    /// little endian bytes, then taking the four most significant bytes (in decreasing order),
+    /// encoding using BASE32 into 8 ASCII characters, and then taking the first 6.
+    ///
+    /// The header `xx/xx/xx` ensures that each directory does not have more than 1024 immediate
+    /// sub-directories.
+    pub fn encode_path(&self, path_buf: &mut PathBuf) {
```
Comment (wupr) [inline src/record/key.rs:242]: I don't feel this function belongs in `impl RemoteId` or even this file. We can have a new `path.rs` file maybe.

**Comment by wupr** [comment]
I think we should make it work without a shell script (I'd love to have at least `autobib path add`). We'd just have to give the user a lot of options about how to name the path, e.g. with your format spec proposal, and make it possible as `autobib path --format <FORMAT_SPEC>`.

File: src/record/key.rs [hunk 2]
```diff
@@ -217,6 +223,44 @@ impl RemoteId {
...
+                unsafe { from_utf8_unchecked(&buffer[0..2]) },
+                unsafe { from_utf8_unchecked(&buffer[2..4]) },
+                unsafe { from_utf8_unchecked(&buffer[4..6]) },
+                &sub_id_encoded,
+                // This appends a `/` or `\`, as platform appropriate, to be clear that this
+                // is a directory
+                "",
+            ]
+            .iter(),
+        );
```
Comment (wupr) [inline src/record/key.rs:262]: The `.iter()` can be omitted, as `IntoIterator` is already implemented for `[T]`

File: src/main.rs [hunk 3]
```diff
@@ -228,6 +245,21 @@ enum Command {
...
+
+// /// The citation key for which to retrieve the path.
+// citation_key: RecordId,
+// /// Show the appropriate directory path.
+// #[arg(short, long, group = "path_action")]
+// show: bool,
+// /// Copy a file into the directory.
+// #[arg(short, long, group = "path_action")]
+// add: Vec<PathBuf>,
+
```
[the remark this answers is not in the document]
Comment (alexrutar) [reply src/main.rs:262]: Yeah exactly, well I didn't mean to leave it commented out in the PR but generally the point...

File: src/main.rs [hunk 4]
```diff
@@ -62,6 +62,10 @@ struct Cli {
     #[arg(short, long, value_name = "PATH", env = "AUTOBIB_DATABASE_PATH")]
     database: Option<PathBuf>,
 
+    /// Use directory for attachments.
+    #[arg(long, value_name = "PATH", env = "AUTOBIB_FILE_DIRECTORY")]
+    file_dir: Option<PathBuf>,
```
[the remark this answers is not in the document]
Comment (alexrutar) [reply src/main.rs:67]: attachment directory? will try to think of some other sensible names

**Comment by alexrutar** [comment]
I implemented a basic `autobib attach` function to take a look at.

My main concern with implementing this here, I suppose, is that the behaviour will be subtly different than the behaviour of `cp` or `mv`. For instance, it is easy to create confusing error messages for instance by using
```
cargo run -- attach arxiv:1212.1873  source.pdf --rename ..
```
which results in a somewhat unclear `error: File exists (os error 17)`. And I do not understand the precise way in which paths are handled by rust so doing something like
```
cargo run -- attach arxiv:1212.1873  source.pdf --rename /a/b
```
results in `error: Read-only file system (os error 30)`. Basically it seems highly unclear what `--rename` is doing.

Maybe the proper implementation is that `--rename` CANNOT take a path argument, but instead is always interpreted as the specific filename. Then we can parse that entry so that it outputs something which is path-friendly, which guarantees that we always insert the file straight into the correct directory, if the user makes the program call.

**Comment by alexrutar** [comment]
I think it is overly complicated to attempt to handle these weird filesystem details. Perhaps the best way to go forwards is to properly implement the `--rename` command so that we have a lot more control over what the file is named to. This is also more justified in my mind since it is providing value which is difficult to implement without access to `autobib` internals.

**Comment by alexrutar** [comment]
Still breaks with
```
cargo run -- attach arxiv:1212.1873 source.pdf --rename a/b/
```
since I cannot figure out how to check for the trailing `/`...

**Comment by alexrutar** [comment]
If we want to add support for automatic opening of attached files, I guess this is sensible to use: https://crates.io/crates/open

**Comment by alexrutar** [comment]
For the format implementation, the parsing is implemented [here](https://docs.rs/biblatex/0.10.0/src/biblatex/types/person.rs.html#12-22). However this is parsed from the `biblatex` crate intermediate representation which processes escapes and such, so I am not sure how much work it is to port the code manually.

**Comment by alexrutar** [comment]
Now using [rapidhash](https://crates.io/crates/rapidhash) instead of highwayhash. Seems to be the most recommended / cutting edge platform-independent non-cryptographic hash function.

File: src/main.rs [hunk 5]
```diff
@@ -432,6 +459,69 @@ fn run_cli(cli: Cli) -> Result<()> {
...
+                if !ext.is_relative() {
+                    bail!("New path must relative");
+                }
+                target.push(ext);
+            } else if let Some(ext) = file.file_name() {
+                target.push(ext);
+            } else {
+                bail!("Source file must not be a directory");
+            }
+
```
Comment (wupr) [inline src/main.rs:510]: A more robust solution is to check whether the target directory is a prefix of the final target path.

**Comment by alexrutar** [comment]
@wupr I think there are still more things to implement but this PR is already quite large and has been open for a while, so I would be happy for it to be merged for now.

Currently, the `autobib path` and `autobib attach` functions have been implemented, and `autobib attach` has the reduced functionality of non-recursive path addition.

I did not implement checking for a trailing backslash and instead just used the Rust default which removes it implicitly. Hopefully this is fine in practice.

File: src/main.rs [hunk 6]
```diff
@@ -719,6 +816,43 @@ fn run_cli(cli: Cli) -> Result<()> {
...
+            let Record { canonical, .. } = record;
+            let mut path = file_dir;
+            canonical.path_hash(&mut path);
+            if mkdir {
+                info!("Creating directory for canonical id '{canonical}'");
+                create_dir_all(&path)?;
+            }
+            println!("{}", path.display());
+            row.commit()?;
+        }
```
Comment (wupr) [inline src/main.rs:855]: This looks identical to the first few lines in the `Command::Attach` branch. Can we extract this?

**Review by wupr — COMMENTED** [review]
Looks great to me, thanks! Just some nitpicking comments.

File: tests/cli.rs [hunk 7]
```diff
@@ -609,6 +618,91 @@ fn info() -> Result<()> {
...
 
+#[test]
+fn test_attach() -> Result<()> {
+    let s = TestState::init()?;
+
+    let source_file = Path::new("tests/resources/path/attachment.txt");
+
+    let temp = assert_fs::NamedTempFile::new("attachment.txt")?;
+    temp.write_file(source_file)?;
+    temp.assert(predicate::path::eq_file(source_file));
```
[the remark this answers is not in the document]
Comment (alexrutar) [reply tests/cli.rs:629]: Haha I am not sure, good point will just write directly to the temp file instead

File: src/path_hash.rs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        path_buf.extend([
+            self.provider(),
+            // SAFETY: BASE32 encoding only returns ASCII bytes
+            unsafe { from_utf8_unchecked(&buffer[0..2]) },
+            unsafe { from_utf8_unchecked(&buffer[2..4]) },
+            unsafe { from_utf8_unchecked(&buffer[4..6]) },
+            &sub_id_encoded,
+            // This appends a `/` or `\`, as platform appropriate, to be clear that this
+            // is a directory
+            "",
```
[the remark this answers is not in the document]
Comment (alexrutar) [reply src/path_hash.rs:46]: It influences the printed text when the user calls `autobib path` to have a trailing slash (so the user knows that this is a directory)

File: src/main.rs [hunk 9]
```diff
@@ -62,17 +64,17 @@ struct Cli {
...
-
     /// Do not require user action.
     ///
     /// This option is set automatically if the standard input is not a terminal.
     #[arg(short = 'I', long, global = true)]
     no_interactive: bool,
-
+    /// Use directory for attachments.
+    #[arg(long, value_name = "PATH", env = "AUTOBIB_ATTACHMENT_DIRECTORY")]
+    attach_dir: Option<PathBuf>,
```
[the remark this answers is not in the document]
Comment (alexrutar) [reply src/main.rs:77]: Done

File: src/main.rs [hunk 10]
```diff
@@ -371,6 +398,9 @@ fn run_cli(cli: Cli) -> Result<()> {
         app_name: env!("CARGO_PKG_NAME").to_owned(),
     })?;
 
+    let data_dir = strategy.data_dir();
+    create_dir_all(&data_dir)?;
```
[the remark this answers is not in the document]
Comment (alexrutar) [reply src/main.rs:402]: I was thinking about doing this - and also making the configuration lazily loaded. But I wanted to wrap this into a single 'filesystem state' struct which does all of this automatically. I will create an issue and fix this in a different PR

[the remark this answers is not in the document]
Comment (alexrutar) [reply src/main.rs:402]: Created issue #146

File: src/path_hash.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (alexrutar) [reply src/path_hash.rs:46]: One possibility is to remove this here, and instead add it immediately before the print call?

File: src/main.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (wupr) [reply src/main.rs:402]: Great!

File: src/path_hash.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (wupr) [reply src/path_hash.rs:46]: I see. I think I would support adding the trailing slash just for the printing.

**Pull request merged**